app/rdf/query-cube-preview.ts

```diff
@@ -0,0 +1,278 @@
+import uniqBy from "lodash/uniqBy";
+import { Quad } from "rdf-js";
+import ParsingClient from "sparql-http-client/ParsingClient";
+
+import {
+  BaseComponent,
+  BaseDimension,
+  DataCubePreview,
+  Dimension,
+  Measure,
+  TemporalDimension,
+} from "@/domain/data";
+import { truthy } from "@/domain/types";
+import { resolveDimensionType, resolveMeasureType } from "@/graphql/resolvers";
+
+import * as ns from "./namespace";
+import {
+  getDataKind,
+  getIsNumerical,
+  getScaleType,
+  getTimeFormat,
+  getTimeUnit,
+  parseNumericalTerm,
+  parseResolution,
+} from "./parse";
+import { buildLocalizedSubQuery } from "./query-utils";
+
+// ?s ?p ?o
+type ShortQuad = Omit<Quad, "subject" | "predicate" | "object"> & {
+  s: Quad["subject"];
+  p: Quad["predicate"];
+  o: Quad["object"];
+};
+
+export const getCubePreview = async (
+  iri: string,
+  options: {
+    locale: string;
+    latest: Boolean;
+    sparqlClient: ParsingClient;
+  }
+): Promise<DataCubePreview> => {
+  const { sparqlClient, locale } = options;
+  const qs: ShortQuad[] = await sparqlClient.query
+    .construct(
+      `PREFIX cube: <https://cube.link/>
+PREFIX meta: <https://cube.link/meta/>
+PREFIX qudt: <http://qudt.org/schema/qudt/>
+PREFIX rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#>
+PREFIX rdfs: <http://www.w3.org/2000/01/rdf-schema#>
+PREFIX schema: <http://schema.org/>
+PREFIX sh: <http://www.w3.org/ns/shacl#>
+PREFIX time: <http://www.w3.org/2006/time#>
+PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>
+    
+CONSTRUCT {
+  ?cube schema:version ?cubeVersion .
+  ?cube cube:observationSet ?observationSet .
+  ?observationSet cube:observation ?observation .
+
+  ?dimension sh:path ?dimensionIri .
+  ?dimension sh:datatype ?dimensionDataType .
+  ?dimension rdf:type ?dimensionType .
+  ?dimension qudt:scaleType ?dimensionScaleType .
+  ?dimension qudt:unit ?dimensionUnit .
+  ?dimensionUnit schema:name ?dimensionUnitLabel .
+  ?dimensionUnit qudt:CurrencyUnit ?dimensionUnitIsCurrency .
+  ?dimensionUnit qudt:currencyExponent ?dimensionUnitCurrencyExponent .
+  ?dimension sh:order ?dimensionOrder .
+  ?dimension meta:dataKind ?dimensionDataKind .
+  ?dimensionDataKind rdf:type ?dimensionDataKindType .
+  ?dimensionDataKind time:unitType ?dimensionTimeUnitType .
+  ?dimension schema:name ?dimensionLabel .
+  ?dimension schema:description ?dimensionDescription .
+
+  ?observation ?observationPredicate ?observationValue .
+  ?observation ?observationPredicate ?observationValueLabel .
+} WHERE {
+  VALUES ?cube { <${iri}> }
+  FILTER(EXISTS { ?cube a cube:Cube . }) {}
```

Comment:
nit: in the current case you do not query for cube metadata within this query - thus you can simply omit the part `{} UNION` as it only adds empty set to the `WHERE` clause.
and I guess `FILTER(EXISTS { ?cube a cube:Cube . })` isn't needed neither ....

Reply:
For some reason omitting this part leads to worse performance (around 0.5s-0.7s vs 0.3s-0.5s when it's there) – as I am not sure why it's like this, and both queries lead to the same results, I think it could be better to keep it 👍